main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala

```diff
@@ -444,6 +444,43 @@ object Unification {
     visit(liftM(fs.head), fs.tail)
   }
 
+  /**
+    * Purifies all effect variables in `tpe` (which should be the type of a letrec-bound function).
```

Comment:
But only the ones at the right level?

Reply:
Good point. I thought about it too, but was unsure what to do. It's is probably correct to consider higher levels, but then again there would have been a call to this function for that level. So maybe not necessary. Only doing it for the same level seemed the safest.